spinalcordtoolbox/scripts/sct_detect_pmj.py

```diff
@@ -182,6 +186,8 @@ def get_max_position(self):
             img_pred_maxValue = np.max(img_pred.data)  # get the max value of the detection map
             self.pa_coord = np.where(img_pred.data == img_pred_maxValue)[0][0]
             self.is_coord = np.where(img_pred.data == img_pred_maxValue)[1][0]
```

Comment:
Also related to https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3457.

Reply:
> Since the PA and IS coordinates are detected based on the mid-sagittal slice, do we want to consider re-running detection on the RL-adjusted sagittal slice?

This is an excellent point indeed. I feel like this goes hand in hand with the aim at getting higher precision in PMJ detection #3457 and likely involves deeper investigations. I am afraid that, by re-estimating PMJ _after_ RL adjustment, the detection will fail in some cases. The detection is based on machine learning and it is not 100% robust. So, if it fails in, let's say 5% of the cases, then doing it twice will increase the probability of failure (in this case, increase the probability to 9.75% considering independent and identically distributed events).

Reply:
My thinking here is that we would use @sandrinebedard's RL sliding window approach as a way to improve the initial sagittal slice selection. So, effectively the change would be this instead:

```python
# Before this PR
extract_mid_sagittal_slice()
detect()

# After this PR
extract_adjusted_sagittal_slice() 
# It would be like we never used the mid-sagittal slice in the first place
# In other words, conceptually, there is still only 1 "main" detection
detect()
```

> So, if it fails in, let's say 5% of the cases, then doing it twice will increase the probability of failure (in this case, increase the probability to 9.75% considering independent and identically distributed events).

I'm not sure the failure rate would compound in the way you describe, because in practice, we wouldn't  _really_ be detecting the PMJ twice; the "first detection" would just be a more involved way of picking a sagittal slice index for the real detection.

Reply:
Ah I see. The thing is, so far, @sandrinebedard was applying `extract_mid_sagittal_slice` based on the detected PMJ (it is required in order to center the sliding window), so if there is no detected PMJ, parameters of the sliding window should probably be revised (eg, making it bigger), and the symmetrization also become more tricky because other tissue will be involved.. anyway, I agree this is something worth investigating.

Reply:
Sorry, I tried to rephrase my comment a bit. There would still be a detected PMJ, but it would be used only to select the ROI.

Reply:
also, about this suggestion:
~~~
# After this PR
extract_adjusted_sagittal_slice()  # contains a detection, but only used to select the ROI for correlation
detect()
~~~
we would need to add another symmetrization step _after_ PMJ detection because the symmetry detection around the PMJ might not be the same as the symmetry in a bigger chunk of an image. Example: if someone has a strong scoliosis, the symmetry point around the PMJ will likely be different

Reply:
My apologies, my suggestion was unclear. I completely agree with everything you've said. I've rewritten my comment to better incorporate the points you've made.

Reply:
no worries! And thank you for this very insightful conversation. Your [updated suggestion](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3468#discussion_r671706822) indeed sounds like a good approach, but I am still hesitant because it does involve two `detect()` steps (hence higher probability for failure). But still worth investigating, I agree.

We should also tackle lower hanging fruits, such as this one:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/96a8b4fe5ee068492ae8579d4f1ae1157a9d8c95/spinalcordtoolbox/scripts/sct_detect_pmj.py#L223

Reply:
I'll investigate @joshuacwnewton 's [suggestion](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3468#discussion_r671706822)!

Reply:
I tested out @joshuacwnewton 's [suggestion ](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3468#discussion_r671706822) and compared with only running correlation after the detection on ` uk-biobank-processed` dataset.

I checked out ~200 images

* The majority of the PMJ detection were very similar with both methods. 
* 16 detections were improved when re-running the detection on the RL-adjusted sagittal slice

    <details>
    <summary>sub-1000710</summary>

    ![pmj_detect_example](https://user-images.githubusercontent.com/71230552/126208952-9009ad00-c109-4fd7-bb19-5a9d097de471.gif)



    </details>

    <details>
    <summary>sub-1001427</summary>

    ![pmj_detect_example2](https://user-images.githubusercontent.com/71230552/126209850-23b45bd3-4692-413a-9edd-62d3d52205f8.gif)



    </details>

* 1 was worst when re-running the detection on the RL-adjusted sagittal slice
    <details>
    <summary>sub-1002552</summary>
    
    ![pmj_detect_example2](https://user-images.githubusercontent.com/71230552/126210391-adb3cdf7-c398-4ed5-976b-9dce5516856e.gif)
    </details>

    

* When detection failed, it was the case for both methods

I would keep re applying `detect()` with R-L adjusted slice! 

The results are in `~/duke/temp/sebeda/pmj_csa/test_pmj_1` and `~/duke/temp/sebeda/pmj_csa/test_pmj_2`
where 1 is for applying joshua's suggestion and 2 for the original method

Reply:
> I tested out @joshuacwnewton 's suggestion and compared with only running correlation after the detection on uk-biobank-processed dataset.

Thank you so much for taking the time to implement and test this! :heart: 

> I checked out ~200 images

Wow! Was this done against ground truth, or just manually comparing by e.g. skimming QC? (Asking as I'm not sure if we have manual PMJ labels for this data or not.)

Reply:
> Wow! Was this done against ground truth, or just visually comparing by e.g. skimming QC? (Asking as I'm not sure if we have manual PMJ labels for this data or not.)

No this was just visually comparing the QC of both methods, I actually  just received the manual labels for this data 😄!